corona-warn-app/cwa-app-android
# Refactor missused ViewModels, support automated UI tests and screenshots (EXPOSUREAPP-2946)

## Why is this PR so huge? :fearful:
Ripping the :adhesive_bandage:  off in one swoop saves us weeks of work.

The core idea is that we want to instantiate screens on their own to automate UI tests and take automated screenshots. This was impossible because there were too many dependencies that would have to be satisfied for each screen, sepperation was not possible because each screen accessed various other components via static access.

We had a lot of :ramen: code based on (`TracingViewModel`, `SubmissionViewModel` and `SettingsViewModel`) that was very interwoven in the apps structure. This PR removes the :ramen: and creates a one-way information flow. By that I mean that information flows from internal components (i.e. Bluetooth state) to external components (Fragments showing something) without extra roundtrips (i.e. we don't want individual fragments calling `refreshBluetooth`).

## What changed
* Reactive information flow for many elements, i.e. UI subscribes data, data changes, UI updates automatically, the UI no longer manually calls refresh in most cases.
* Fewer singleton access to allow easier testing.
* Most `***FormaterHelper` calls have been removed from the layout files, we had too much logic happening in XML files. Too difficult to add new features, too easy to break existing logic.
* We now have UI "state" objects. The viewmodels have a `Flow<State>` that can be observed, and if it changes the Fragment will update the layout via databinding. The remaining databinding code just sets the state variables to the relevant views. So now we have the best of both worlds. Not too complex logic in XML code. No extra boilerplate code in Fragments.
    * The states can easily be unit-tested on their own.
    * When writing instrumentation tests or doing  automated screenshots, we only need to create mock state objects to make the UI look a specific way.

The above patterns are something we should use throughout the app. This PR already refactors a lot but does not refactor everything. Only the places that were connected to the classes that needed to be refactored in relation to the `HomeFragment`... which is a lot :sweat_smile: 

## Testing
* Check the notification settings behavior in different states
* Check the tracing settings behavior in different states
* Check the behavior of the tracing status headline on the home screen in different states
* Check the risk card behavior in different states
* Check the submission card behavior in different states
* Check the risk details screen in different states
* Check submission flow
